pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java

```diff
@@ -480,6 +487,20 @@ public void close() throws PulsarServerException {
         }
     }
 
+    private CompletableFuture<Void> addTimeoutHandling(CompletableFuture<Void> future) {
+        ScheduledExecutorService shutdownExecutor = Executors.newSingleThreadScheduledExecutor(
+                new DefaultThreadFactory(getClass().getSimpleName() + "-shutdown"));
+        FutureUtil.addTimeoutHandling(future,
+                Duration.ofMillis(Math.max(1L, getConfiguration().getBrokerShutdownTimeoutMs())),
+                shutdownExecutor, () -> FutureUtil.createTimeoutException("Timeout in close", getClass(), "close"));
+        future.handle((v, t) -> {
+            // shutdown the shutdown executor
+            shutdownExecutor.shutdown();
```

Comment:
`shutdown` is called from the thread executing in the shutdownExecutor. Therefore calling `.shutdownDown` or `.await` aren't suitable. (It's not really needed either since there aren't any blocking methods that the shutdownExecutor calls.)

Reply:
After thinking about it again, yes. I'll rewrite the logic to use `awaitTermination`.

Reply:
I have now rewritten the logic to use `awaitTermination`. @merlimat please take another look

Reply:
oh sorry, I didn't look at the full context of the code where this was. In this case, I'll simply use `shutdownNow` to terminate the scheduler used to schedule the timeout handling for the CompletableFuture.

Reply:
Switched to use `shutdownNow`